File: db4o.net/Db4objects.Db4o/Db4objects.Db4o/Foundation/NonblockingQueue.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using System;
using System.Collections;
using Db4objects.Db4o.Foundation;

namespace Db4objects.Db4o.Foundation
{
	/// <summary>Unbounded queue.</summary>
	/// <remarks>Unbounded queue.</remarks>
	/// <exclude></exclude>
	public class NonblockingQueue : IQueue4
	{
		private List4 _insertionPoint;

		private List4 _next;

		public void Add(object obj)
		{
			List4 newNode = new List4(null, obj);
			if (_insertionPoint == null)
			{
				_next = newNode;
			}
			else
			{
				_insertionPoint._next = newNode;
			}
			_insertionPoint = newNode;
		}

		public object Next()
		{
			if (_next == null)
			{
				return null;
			}
			object ret = ((object)_next._element);
			RemoveNext();
			return ret;
		}

		private void RemoveNext()
		{
			_next = ((List4)_next._next);
			if (_next == null)
			{
				_insertionPoint = null;
			}
		}

		public virtual object NextMatching(IPredicate4 condition)
		{
			if (null == condition)
			{
				throw new ArgumentNullException();
			}
			List4 current = _next;
			List4 previous = null;
			while (null != current)
			{
				object element = ((object)current._element);
				if (condition.Match(element))
				{
					if (previous == null)
					{
						RemoveNext();
					}
					else
					{
						previous._next = ((List4)current._next);
					}
					return element;
				}
				previous = current;
				current = ((List4)current._next);
			}
			return null;
		}

		public bool HasNext()
		{
			return _next != null;
		}

		public virtual IEnumerator Iterator()
		{
			List4 origInsertionPoint = _insertionPoint;
			List4 origNext = _next;
			return new _Iterator4Impl_82(this, origInsertionPoint, origNext, _next);
		}

		private sealed class _Iterator4Impl_82 : Iterator4Impl
		{
			public _Iterator4Impl_82(NonblockingQueue _enclosing, List4 origInsertionPoint, List4
				 origNext, List4 baseArg1) : base(baseArg1)
			{
				this._enclosing = _enclosing;
				this.origInsertionPoint = origInsertionPoint;
				this.origNext = origNext;
			}

			public override bool MoveNext()
			{
				if (origInsertionPoint != this._enclosing._insertionPoint || origNext != this._enclosing
					._next)
				{
					throw new InvalidOperationException();
				}
				return base.MoveNext();
			}

			private readonly NonblockingQueue _enclosing;

			private readonly List4 origInsertionPoint;

			private readonly List4 origNext;
		}
	}
}

File: db4oj/core/src/com/db4o/foundation/NonblockingQueue.java
/* Copyright (C) 2004 -2007  Versant Inc.   http://www.db4o.com */

package com.db4o.foundation;

/**
 * Unbounded queue.
 * @exclude
 */
public class NonblockingQueue<T> implements Queue4<T> {

	private List4<T> _insertionPoint;
	private List4<T> _next;
	
    /* (non-Javadoc)
	 * @see com.db4o.foundation.Queue4#add(java.lang.Object)
	 */
    public final void add(T obj) {
    	List4<T> newNode = new List4<T>(null, obj);
    	if (_insertionPoint == null) {
    		_next = newNode;
    	} else {
    		_insertionPoint._next = newNode;
    	}
    	_insertionPoint = newNode;
    }
    
	/* (non-Javadoc)
	 * @see com.db4o.foundation.Queue4#next()
	 */
	public final T next() {
		if(_next == null){
			return null;
		}
		T ret = _next._element;
		removeNext();
		return ret;
	}

	private void removeNext() {
		_next = _next._next;
		if (_next == null) {
			_insertionPoint = null;
		}
	}
	
	public T nextMatching(Predicate4<T> condition) {
		if (null == condition) {
			throw new ArgumentNullException();
		}
		
		List4<T> current = _next;
		List4<T> previous = null;
		while (null != current) {
			final T element = current._element;
			if (condition.match(element)) {
				if (previous == null) {
					removeNext();
				} else {
					previous._next = current._next;
				}
				return element;
			}
			previous = current;
			current = current._next;
		}
		return null;
	}
    
    /* (non-Javadoc)
	 * @see com.db4o.foundation.Queue4#hasNext()
	 */
    public final boolean hasNext() {
        return _next != null;
    }

	/* (non-Javadoc)
	 * @see com.db4o.foundation.Queue4#iterator()
	 */
	public Iterator4<T> iterator() {
		final List4<T> origInsertionPoint = _insertionPoint;
		final List4<T> origNext = _next;
		return new Iterator4Impl<T>(_next) {
			
			public boolean moveNext() {
				if (origInsertionPoint != _insertionPoint || origNext != _next) {
					throw new IllegalStateException();
				}
				return super.moveNext();
			}
		};
	}
}
